gpu/containers/include/pcl/gpu/containers/device_array.h

```diff
@@ -36,233 +36,253 @@
-
-            /** \brief Downloads data from internal buffer to CPU memory
-               * \param data host vector to download to                 
-               * \param cols Output stride in elements between two consecutive rows for host vector.
-               * */
-            template<class A>
-            void download(std::vector<T, A>& data, int& cols) const;
-                                      
-            /** \brief Returns pointer to given row in internal buffer. 
-              * \param y row index   
-              * */
-            T* ptr(int y = 0);             
-
-            /** \brief Returns const pointer to given row in internal buffer. 
-              * \param y row index   
-              * */
-            const T* ptr(int y = 0) const;            
-            
-            //using DeviceMemory2D::ptr;            
-
-            /** \brief Returns pointer for internal buffer in GPU memory. */
-            operator T*();
-
-            /** \brief Returns const pointer for internal buffer in GPU memory. */
-            operator const T*() const;                        
-            
-            /** \brief Returns number of elements in each row. */
-            int cols() const;
-
-            /** \brief Returns number of rows. */
-            int rows() const;
-
-            /** \brief Returns step in elements. */
-            std::size_t elem_step() const;
-        };        
-    }
-
-    namespace device
-    {
-        using pcl::gpu::DeviceArray;
-        using pcl::gpu::DeviceArray2D;
-    }
-}
+namespace pcl {
+namespace gpu {
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+/** \brief @b DeviceArray class
+ *
+ * \note Typed container for GPU memory with reference counting.
+ *
+ * \author Anatoly Baksheev
+ */
+template <class T>
+class PCL_EXPORTS DeviceArray : public DeviceMemory {
+public:
+  /** \brief Element type. */
+  using type = T;
+
+  /** \brief Element size. */
+  enum { elem_size = sizeof(T) };
+
+  /** \brief Empty constructor. */
+  DeviceArray();
+
+  /** \brief Allocates internal buffer in GPU memory
+   * \param size number of elements to allocate
+   * */
+  DeviceArray(std::size_t size);
+
+  /** \brief Initializes with user allocated buffer. Reference counting is disabled in
+   * this case. \param ptr pointer to buffer \param size elements number
+   * */
+  DeviceArray(T* ptr, std::size_t size);
+
+  /** \brief Copy constructor. Just increments reference counter. */
+  DeviceArray(const DeviceArray& other);
+
+  /** \brief Assignment operator. Just increments reference counter. */
+  DeviceArray&
+  operator=(const DeviceArray& other);
+
+  /** \brief Allocates internal buffer in GPU memory. If internal buffer was created
+   * before the function recreates it with new size. If new and old sizes are equal it
+   * does nothing. \param size elements number
```

Comment:
Please search for doxygen issues using a regex like ` +\* *.+\\` (some spaces, then a star, a space, some chars and then a backslash)
`\param` should come on a new line

Reply:
Thanks for the comment! Can you kindly explain to me how you spotted the doxygen warnings? I built the documentation with `make doc` and raises warnings about not documented parameters by setting `WARN_NO_PARAMDOC = YES` in the `doxyfile.in`. Nevertheless, I could not identify the errors you spotted related to the `DeviceArray<T>`. I want to identify the warnings before my next commit, and thus would love to learn how you spotted the errors!

Reply:
I saw it while browsing the patch. For checking, you can use the following regex in vim:
` \+\* \+.\+\\` (ignoring all the slash it is basically space followed by *, followed by spaces and then some letters and then a backslash. PCRE regex for the same is `[ *]\*+ +.+\\`)

You can use rg or grep or similar tool for finding all such occurances

Reply:
Thanks for the tip! I fixed many issues and hope not to have missed any. Do you still find ill-formatted code?